workers/ohsome_quality_analyst/indicators/mapping_saturation/fit.py

```diff
@@ -1,111 +1,71 @@
-"""Fit sigmoid functions to data.
+"""Run all models and choose best fit based on Mean Absolute Error."""
 
-Nonlinear least squares optimization is used to fit data to sigmoid functions which
-include up to four jumps.
-
-Best fitting function is chosen by the applying mean-squared error criterion.
-"""
 
 import logging
 from dataclasses import dataclass
-from typing import Tuple
+from typing import List
 
-import numpy as np
-from scipy.optimize import curve_fit
+from numpy import float64, ndarray
+from rpy2.rinterface_lib.embedded import RRuntimeError
 
 from ohsome_quality_analyst.indicators.mapping_saturation import metrics, models
 
 
 @dataclass
-class Fit:
-    model_name: str
-    ydata: np.ndarray
+class FittedModel:
+    asymptote: float64
+    coefficients: dict
+    function_formula: str
+    metric: float64
     metric_name: str
-    metric: np.float64
-    asymptote: np.float64
-
+    model_name: str
+    fitted_values: ndarray
 
-def get_best_fit(xdata: np.ndarray, ydata: np.ndarray) -> Fit:
-    """Get best fit function based on Mean Squared Error.
 
-    Fit sigmoid_1 to sigmoid_4 to given data and return best fit based on Mean Squared
-    Error.
-    """
-    best_fit = None
+def run_all_models(xdata: ndarray, ydata: ndarray) -> List[FittedModel]:
     # TODO: Run for every model
-    model = models.Sigmoid()
-    p0 = model.get_initial_guess(xdata, ydata)
-    bounds = model.get_bounds(xdata, ydata)
-    try:
-        # curve_fit: Use non-linear least squares to fit a function, f, to data.
-        # popt: Optimal values for the parameters as array
-        # pcov: The estimated covariance of popt as 2-D array
-        popt, pcov = curve_fit(
-            model.function, xdata=xdata, ydata=ydata, p0=p0, bounds=bounds
+    fitted_models = []
+    for model in (
+        models.Sigmoid(),
+        models.SSlogis(),
+        models.SSdoubleS(),
+        models.SSfpl(),
+        models.SSasymp(),
+        models.SSmicmen(),
+    ):
+        logging.info("Run {}".format(model.name))
+        try:
+            coef = model.fit(xdata, ydata)
+        except RRuntimeError as error:
```

Comment:
you might need to change this to `except (RRuntimeError, RuntimeError) as error:`

I tested it locally and got an error that was initially not catched by your line.

```
2021-12-09 17:40:18,791 - WARNING - callbacks.py - consolewrite_warnerror - R[write to console]: Error in nls(y ~ cbind(1, 1/(1 + exp((xmid - x)/exp(lscal)))), data = xy,  : 
  number of iterations exceeded maximum of 50

2021-12-09 17:40:18,792 - WARNING - fit.py - run_all_models - Could not run model Self-Starting Nls Four-Parameter Logistic Model due to RRuntimeError: Error in nls(y ~ cbind(1, 1/(1 + exp((xmid - x)/exp(lscal)))), data = xy,  : 
  number of iterations exceeded maximum of 50
```

Reply:
`Could not run model Self-Starting Nls Four-Parameter Logistic Model due to RRuntimeError` is logged by you manually.

Reply:
The first of the two error message which you shared above is actually just a logging message. Before the `RRuntimeError` is raised the rpy2 library will log the message and then raise the error. The first of the two logging messages above is from rpy2 then the error gets raised which is caught by OQT and error messages is logged again, resulting in two logs with the same message. Does this make sense?